RPI5: add scmi and wifi support

GrygiriiS opened the pull request:
**Required**: PR #https://github.com/xen-troops/meta-xt-rpi5/pull/37

- Switch ARM TF and XEN on rpi5 dev branches. 
  The ARM TF SCMI already merged, but I do expect some changes during full integration and testing.
  The XEN still in progress.
- added DT overlays for both SCMI and WIFI.
- updated u-boot script
- Updated rpi5.yaml to enable SCMI and add option to Enable WIFI

NOTE1.  AUTOREV still in use, but please do NOT concentrate to much attention on it. It will be fixed.
NOTE2. Private branches are used in rpi5.yaml - it will be fixed.
NOTE3. Without above 2 notes it terribly hard to put it all together for integration and testing.

It was tested with (boot and wifi):

-  a bit modified https://github.com/xen-troops/xen/pull/209
- current RPI5 ARM TF  https://github.com/xen-troops/arm-trusted-firmware/tree/rpi5_dev
- Zephyr SCMI https://github.com/xen-troops/zephyr/pull/124
- Zephyr DOM0 https://github.com/xen-troops/zephyr-dom0-xt/pull/36

rshym commented on xt-prod-devel-rpi5-domd/recipes-kernel/linux/linux-raspberrypi_6.6.bbappend:
```diff
@@ -4,15 +4,18 @@ FILESEXTRAPATHS:prepend := "${THISDIR}/files:"
 SRC_URI:append = " \
     file://optee.cfg \
 "
-
```
Looks like a random change.

rshym commented on rpi5.yaml:
```diff
@@ -312,6 +312,8 @@ images:
           "%{XEN_DT_SCMI_NAME}.dtbo": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/%{XEN_DT_SCMI_NAME}.dtbo"
           "%{SOC_FAMILY}-%{MACHINE}-domd-scmi.dtbo": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/%{SOC_FAMILY}-%{MACHINE}-domd-scmi.dtbo"
           "%{SOC_FAMILY}-%{MACHINE}-pcie1-scmi.dtbo": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/%{SOC_FAMILY}-%{MACHINE}-pcie1-scmi.dtbo"
+          "%{SOC_FAMILY}-%{MACHINE}-wifi.dtbo": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/%{SOC_FAMILY}-%{MACHINE}-wifi.dtbo"
+          "wifi-passthrough.dtbo": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/wifi-passthrough.dtbo"
```
Seems you should put this under the `--ENABLE_WIFI` option.
```
images:
  full:
    partitions:
      boot:
      ....

rshym commented on rpi5.yaml:
```diff
@@ -50,8 +50,8 @@ common_data:
     - type: git
-      url: "https://github.com/xen-troops/meta-xt-prod-devel-rpi5.git"
-      rev: "main"
+      url: "https://github.com/GrygiriiS/meta-xt-prod-devel-rpi5.git"
+      rev: "rpi5_scmi_wifi"
```
This change should be in PR.

rshym commented on rpi5.yaml:
```diff
@@ -68,8 +68,8 @@ components:
       - type: west
-        url: "https://github.com/xen-troops/zephyr-dom0-xt.git"
-        rev: "rpi5_dom0_dev"
+        url: "https://github.com/GrygiriiS/zephyr-dom0-xt.git"
+        rev: "rpi5_dom0_dev_scmi"
```
This change should be in PR.

rshym commented on rpi5.yaml:
```diff
@@ -142,8 +142,8 @@ components:
       - type: git
-        url: "https://github.com/xen-troops/meta-xt-rpi5.git"
-        rev: "main"
+        url: "https://github.com/GrygiriiS/meta-xt-rpi5.git"
+        rev: "rpi5_scmi_wifi"
```
This change should be in PR.

rshym commented:
`Acked-by: Ruslan Shymkevych <ruslan_shymkevych@epam.com>`

GrygiriiS commented:
This PR updated:
- dropped private dev branches
- dropped Zephyr Dom0 Scmi
In general, it's ready, but we'ra waiting for Xen SCMI support to be finished.


Test branch to be used for integration can be found here:
https://github.com/GrygiriiS/meta-xt-prod-devel-rpi5/tree/rpi5_scmi_wifi_v1_dev

oleksiimoisieiev commented:
I think README should be updated as well

rshym commented:
After the repush, the first commit message still mentions AUTOREV.
The second commit has 'No end-of-the-file" (Red crossed circle).

lorc commented on xt-prod-devel-rpi5-domd/recipes-extended/xen/files/xen-scmi.cfg:
```diff
@@ -0,0 +1,2 @@
+CONFIG_ARM_SCI=y
```
Is this change should be in this  commit? Commit message does not mention enabling SCMI

lorc commented on xt-prod-devel-rpi5-domd/recipes-bsp/trusted-firmware-a/trusted-firmware-a_git.bbappend:
```diff
@@ -0,0 +1,13 @@
+# Current master branch
+SRC_URI_TRUSTED_FIRMWARE_A = "git://github.com/xen-troops/arm-trusted-firmware.git;protocol=https"
+
+SRCREV_tfa = "${AUTOREV}"
```
As we are going to merge this into the main branch, we don't need AUTOREV here

lorc commented on xt-prod-devel-rpi5-domd/recipes-kernel/linux/files/bcm2712-raspberrypi5-xen-scmi.dtso:
```diff
@@ -0,0 +1,150 @@
+// SPDX-License-Identifier: GPL-2.0
```
It would be nice to have more extended commit message for this change

lorc commented on xt-prod-devel-rpi5-domd/recipes-bsp/xt-rpi-u-boot-scr/xt-rpi-u-boot-scr.bbappend:
```diff
@@ -1,8 +1,11 @@
 DOMD_OVERLAYS:append = "${@bb.utils.contains("MACHINE_FEATURES", "scmi", " ${SCMI_DOMD_DT_NAME}.dtbo", "", d)}"
 DOMD_OVERLAYS:append = "${@bb.utils.contains("MACHINE_FEATURES", "scmi", " ${SCMI_PCIE1_DT_NAME_ADD}", "", d)}"
+
+DOMD_OVERLAYS:append = "${@bb.utils.contains("MACHINE_FEATURES", "domd_wifi", " ${DOMD_WIFI_DT_NAME}.dtbo", "", d)}"
```
IMO, this approach for adding dtbo is getting out of hand. Lots of almost duplicated code....

Also, commit message clearly mentions `linux:` but you are making changes not only to linux.

lorc commented on rpi5.yaml:
```diff
@@ -365,3 +365,23 @@ parameters:
                 - [DOMD_OVERLAYS, "%{SOC_FAMILY}-%{MACHINE}-pcie1.dtbo"]
                 - [MACHINE_FEATURES:append, " domd_nvme"]
+
+  ENABLE_WIFI:
```
IMO, this patch can be folded into previous one

lorc commented on rpi5.yaml:
```diff
@@ -271,25 +271,7 @@ images:
-          "start4db.elf": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4db.elf"
-          "start4cd.elf": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4cd.elf"
-          "start4x.elf":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/start4x.elf"
+          "/": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles"
```
Have you tested this? Does ninja correctly handles changes in the files?

GrygiriiS replied:
ninja has nothing to do with this - this is handled by **Rouge** tool. And it works for directory.

lorc replied:
`ninja` has everything to do with this, as it parses dependencies and decides what should be done before invoking the `rouge` tool. With your approach there is no guarantee that image file will include any bootfiles at all. If Yocto for some reason will not build them, rouge will pack just an empty directory. Also, ninja might not trigger image rebuild if some files inside directory were changed. We need to test this.

GrygiriiS replied:
Have you tried by your self delete any file deploy/images/? run **ninja** it will succeed, but **ninja full.img** - will fail. So do not know what "ninja handles dependency" function you refer here to :(

Next in "build.ninja" there is below - which is just call of **rouge** with parameters:
build full.img: rouge yocto/build-domd/tmp/deploy/images/raspberrypi5/armstub8-2712.bin $
    yocto/build-domd/tmp/deploy/images/raspberrypi5/bootfiles $
    yocto/build-domd/tmp/deploy/images/raspberrypi5/Image.gz $


And the last, and most important, we DO NOT maintain **bootfiles** recipe. it come from **meta-raspberrypi** and can be changed there by adding or removing some bootloader files. Now if smth is removed - ok **rouge** will fail. But if added - the build will succeed (including image generation), but RPI5 boot will fail. It make no sense for us to maintain full list of RPI5 bootfiles here (and moreover It is not safe) as we do not own it. What is reasonable is to have dependency on one or two files and if it's satisfied then assume **bootfiles** build is ok and just copy ALL bootfiles in the rootfs image.

GrygiriiS replied:
And moulin documentation:
https://moulin.readthedocs.io/en/latest/user-reference.html
 yocto builder https://moulin.readthedocs.io/en/latest/user-reference.html#yocto-builder

`target_images - list of image files that should be generated by this component as a result of invoking $ bitbake {build_target}. Every component should generate at least one image file.
`
Note. **Every component should generate at least one image file.** So no need to list ALL binaries.

`additional_deps - list of additional dependencies. This is basically target_images produced by other components. You can use those to implement build dependencies between components. For example, if your system needs to have DomU’s kernel image on Dom0 file system, you might want to add path to DomU’s kernel into additional_deps of Dom0’s config. This will ensure that Dom0 will be built after DomU.`

Note. **additional_deps** description clearly mention that NO NEED to add dependencies for ALL images, just  for key images.

Rouge docs: https://moulin.readthedocs.io/en/latest/rouge.html#filesystem-image-with-files   
`
You can specify parent folders for remote and these folders will be created on the destination filesystem. You may specify not only files but directories also. If the local directory contains subdirectories, they will be created unders the remote directory. 
`
As per above, just again. this change is valid usage of moulin/rouge.

lorc replied:
>  Now if smth is removed - ok rouge will fail. But if added - the build will succeed (including image generation), but RPI5 boot will fail. It make no sense for us to maintain full list of RPI5 bootfiles here (and moreover It is not safe) as we do not own it. 

In my opinion it is better to fail during the build process, not during runtime. This will save us a lot of headache in case of any issues in the future. This is another reason why I want to maintain the complete list of required files.

oleksiimoisieiev commented on xt-prod-devel-rpi5-domd/recipes-extended/xen/xen-source.inc:
```diff
@@ -0,0 +1,5 @@
+
+XEN_URL = "git://github.com/xen-troops/xen.git;protocol=https;branch=rpi5_dev"
+XEN_REL = "4.19"
+XEN_REV = "2214507e72c2c2217b3234478912be6581324fbf"
```
I think AUTOREV should be set here.

oleksiimoisieiev commented on rpi5.yaml:
```diff
@@ -271,25 +271,7 @@ images:
           "armstub8-2712.bin": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/armstub8-2712.bin"
-          "bootcode.bin": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/bootcode.bin"
-          "config.txt":   "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/config.txt"
-          "cmdline.txt":  "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/bootfiles/cmdline.txt"
```
This could be the fix of https://github.com/xen-troops/meta-xt-prod-devel-rpi5/issues/3

GrygiriiS replied:
I have no idea

oleksiimoisieiev commented on rpi5.yaml:
```diff
@@ -271,25 +271,7 @@ images:
         size: 512 MiB
         items:
           "armstub8-2712.bin": "%{YOCTOS_WORK_DIR}/%{DOMD_BUILD_DIR}/tmp/deploy/images/%{MACHINE}/armstub8-2712.bin"
```
full.imag -> full.img in the commit descrition.

oleksiimoisieiev commented:
This approach to add all scmi related stuff in meta-xt-prod-devel looks as the only acceptable solution. That's because there is a significant difference in Zephyr (Dom0) memory mapping with/w-o SCMI. So we can't make it completely optional (controlled by config param) without making 2 zephyr branches or using special snippets. In any case - every configuration will have it's specific memory-mapping. So when all configuration is on the top layer - it can be used as the reference for making project-specific top meta-layer with or without SCMI enabled.
This approach looks good so 
`Reviewed-by: Oleksii Moisieiev <oleksii_moisieiev@epam.com>`
with a small comments.

lorc commented:
Thanks

`Acked-by: Volodymyr Babchuk <volodymyr_babchuk@epam.com>`
